x-template.js

```diff
@@ -477,40 +370,225 @@ class TemplateEngine {
+      const item = arrayState.map.get(id);
+      TemplateEngine.#removeThrough(item.startNode, node);
+      arrayState.map.delete(id);
+    }
+  }
+
+  // Loops over given array of “entries” to manage an array of nodes.
+  static #commitContentArrayEntries(node, startNode, entries) {
+    const arrayState = TemplateEngine.#getState(startNode, TemplateEngine.#ARRAY_STATE);
+    if (!arrayState.map) {
+      // There is no mapping in our state — create an empty one as our base.
+      TemplateEngine.#clearObject(arrayState);
+      arrayState.map = new Map();
+    }
+
+    // A mapping has already been created — we need to update the items.
+    const ids = new Set(); // Populated in “parseListValue”.
+    let index = 0;
+    for (const entry of entries) {
+      const [id, rawResult] = TemplateEngine.#parseArrayEntry(entry, index, ids);
+      let item = arrayState.map.get(id);
+      if (item) {
+        if (!TemplateEngine.#canReuseDom(item.preparedResult, rawResult)) {
+          TemplateEngine.#recreateArrayItem(item, rawResult);
+        } else {
+          TemplateEngine.#update(item.preparedResult, rawResult);
+        }
       } else {
-        TemplateEngine.#update(state.preparedResult, rawResult);
+        item = TemplateEngine.#createArrayItem(node, id, rawResult);
+        arrayState.map.set(id, item);
       }
-    } else if (category === 'array' || category === 'map') {
-      TemplateEngine.#list(node, startNode, value, category);
-    } else if (category === 'fragment') {
-      if (value.childElementCount === 0) {
-        throw new Error(`Unexpected child element count of zero for given DocumentFragment.`);
+      index++;
+    }
+    for (const [id, item] of arrayState.map.entries()) {
+      if (!ids.has(id)) {
+        TemplateEngine.#removeThrough(item.startNode, item.node);
+        arrayState.map.delete(id);
       }
-      const previousSibling = node.previousSibling;
-      if (previousSibling !== startNode) {
-        TemplateEngine.#removeBetween(startNode, node);
+    }
+    let lastItem;
+    for (const id of ids) {
+      const item = arrayState.map.get(id);
+      // TODO: We should be able to make the following code more performant.
+      const referenceNode = lastItem ? lastItem.node.nextSibling : startNode.nextSibling;
+      if (referenceNode !== item.startNode) {
+        const nodesToMove = [item.startNode];
+        while (nodesToMove[nodesToMove.length - 1] !== item.node) {
+          nodesToMove.push(nodesToMove[nodesToMove.length - 1].nextSibling);
+        }
+        TemplateEngine.#insertAllBefore(referenceNode.parentNode, referenceNode, nodesToMove);
       }
-      node.parentNode.insertBefore(value, node);
-    } else {
+      lastItem = item;
+    }
+  }
+
+  // TODO: #254: Future state where the “moveBefore” API is better-supported.
+  // // Loops over given array of “entries” to manage an array of nodes.
+  // static #commitContentArrayEntries(node, startNode, entries) {
```

Comment:
Eventually (perhaps soon?), we’ll just uncomment this and delete the above handler. This pattern of encoding _what’s next_ has been quite helpful (assuming these comments don’t linger indefinitely).

Reply:
My experience with trying to do that has just been stale PRs. I’m going to leave this as-is for now, but the goal will be to take up all these commented-out code blocks _ahead_ of the `2.x` release.

So, if it turns out we really cannot take this on ahead of `2.x` I’ll likely just paste this snippet into #254 or something for tracking 👌